programs/utilities/block_counter.sc

```diff
@@ -1,28 +1,151 @@
+//Counts blocks in a given region, optionally filtering by block type or tag. 
+//Prints the output into a chat or a file.
+//If vertical scan is set to true, it will return a separate report for each y value.
+//By gnembon and Firigion
+
 __config() ->
 {
    'commands' -> {
+      '' -> 'help',
       '<from_pos> <to_pos>' -> 'count_blocks',
       '<from_pos> <to_pos> <filter>' -> 'count_blocks_filter',
-   }
+      '<from_pos> <to_pos> <blockpredicate>' -> 'count_blocks_predicate',
+      'save' -> ['save_scan', null],
+      'save <name>' -> 'save_scan',
+      'autosave <bool>' -> 'autosave',
+      'layered_scan <bool>' -> 'vertical_scan', // I'm keeping this as an option because doing the vertically spearated scans is 30% to 100% slower
+      'histogram <block>' -> 'histogram',
+      'tally_level <int>' -> 'tally_level',
+      //'book' -> 'get_book_report',
+      'last' -> 'last_report',
+   },
+   'allow_command_conflicts' -> true
 };
 
+
+/////////////////////////////
+///// Scans
+/////////////////////////////
+
 count_blocks(from_pos, to_pos) ->
 (
-    stats = {};
-    volume(from_pos, to_pos, stats:str(_) += 1 );
-    tally(stats, sum(... values(stats)));
+    global_data = make_data_stuct(from_pos, to_pos, null); //make global empty data structure
+    stats = global_data:'stats'; //save reference for easier access
+
+    if(global_vertscan,
+      // do a scan separating results by y value
+      area = global_data:'area';
+      [miny, maxy] = get_minmax_pos(from_pos, to_pos);
+      total = 0;
+
+      volume(from_pos, to_pos, 
+         stats:(miny+floor(total/area)):str(_) += 1; //very slightly faster than calling pos(_):1
+         total += 1; 
+      );
+      vertical_report(),
+
+      // do a regular scan
+      volume(from_pos, to_pos, stats:str(_) += 1 );
+      tally(global_data:'stats', global_data:'volume');
+      global_data:'total' = sum(... values(global_data:'stats'));
```

Comment:
If you're counting all the blocks here, instead of summing all the items of the list couldn't you just do dx * dy * dx?
Cos you have `from_pos` and `to_pos`, so the total is: `abs((from_pos:0 - to_pos:0)*(from_pos:1 - to_pos:1)*(from_pos:2 - to_pos:2))`